osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs

```diff
@@ -55,30 +61,243 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+
+            judgements[0] = Math.Floor(16 * windowMultiplier);
+            judgements[1] = Math.Floor((64 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[2] = Math.Floor((97 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[3] = Math.Floor((127 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[4] = Math.Floor((151 - 3 * overallDifficulty) * windowMultiplier);
+
+            return judgements;
+        }
+
+        private double[] getLazerHitWindows(ScoreInfo score, ManiaDifficultyAttributes attributes)
+        {
+            double[] hitWindows = new double[5];
+
+            // Create a new track of arbitrary length
+            var track = new TrackVirtual(10000);
+            // Apply the total rate change of every mod to the track (i.e. DT = 1.01-2x, HT = 0.5-0.99x)
+            score.Mods.OfType<IApplicableToTrack>().ForEach(m => m.ApplyToTrack(track));
+            // The final clock rate is the rate of the track
+            double clockRate = track.Rate;
+
+            double windowMultiplier = 1 / clockRate;
+
+            if (score.Mods.Any(m => m is ModHardRock))
+                windowMultiplier *= 1 / 1.4;
+            else if (score.Mods.Any(m => m is ModEasy))
+                windowMultiplier *= 1.4;
+
+            if (attributes.OverallDifficulty < 5)
+                hitWindows[0] = (22.4 - 0.6 * attributes.OverallDifficulty) * windowMultiplier;
+            else
+                hitWindows[0] = (24.9 - 1.1 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[1] = (64 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[2] = (97 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[3] = (127 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[4] = (151 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+
+            return hitWindows;
+        }
+
+        private double logPcNote(double x, double deviation) => logErfcApprox(x / (deviation * Math.Sqrt(2)));
+
+        // For values of x above 5, there is a very simple approximation to increase how far you can calculate x.
+        private double logErfcApprox(double x) => x <= 5 ? Math.Log(SpecialFunctions.Erfc(x)) : -Math.Pow(x, 2) - Math.Log(x) - Math.Log(Math.Sqrt(Math.PI));
+
+        // Legacy LN tails take the absolute error of both hit judgements on an LN, so we use a folded normal distribution to calculate it.
+        private double logPcHoldTail(double x, double deviation) => Math.Log(1 - Math.Pow(2 * Normal.CDF(0, deviation * Math.Sqrt(2), x) - 1, 2));
+
+        // Log rules make addition and subtraction of the non-log value non-trivial, these methods simply add and subtract the base value of logs.
+        private double logSum(double l1, double l2)
+        {
+            double maxVal = Math.Max(l1, l2);
+            double minVal = Math.Min(l1, l2);
```

Comment:
Maybe a nitpick, but variables `l1` and `l2` look like numerical values, which doesn't help with code readability. Same in `logDiff` below.

Reply:
Changed to `firstLog` and `secondLog`